### PR TITLE
Enhanced micro_http implementation

### LINKED ISSUE #1191: Enhance micro_http implementation
The current version of `micro_http` supports only the GET method, but we want it to also support PUT and PATCH. Aside from this, we also want to add specific header support and refactor the parsing.

### PR DESCRIPTION
**Pull request by georgepisaltu**
The current version of `micro_http` now supports the GET, PUT and
PATCH methods. Aside from this, we also added specific header
support and refactored the parsing.

Issue #, if available: fixes #1191

Signed-off-by: George Pisaltu <gpl@amazon.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: micro_http/src/common/headers.rs [hunk 1]
```diff
@@ -13,39 +17,178 @@ pub enum Header {
...
     fn raw(&self) -> &'static [u8] {
         match self {
             Header::ContentLength => b"Content-Length",
             Header::ContentType => b"Content-Type",
+            Header::Expect => b"Expect",
+            Header::TransferEncoding => b"Transfer-Encoding",
+        }
+    }
+
+    fn try_from(string: &[u8]) -> Result<Self, RequestError> {
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: Since now we also parse HTTP headers, you should update the documentation in `micro_http/src/lib.rs` and add the supported ones.

File: micro_http/src/common/headers.rs [hunk 2]
```diff
@@ -3,8 +3,12 @@
 
 use std::collections::HashMap;
 use std::io::{Error as WriteError, Write};
+use std::result::Result;
 
 use ascii::{COLON, CR, LF, SP};
+use std::cell::RefCell;
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: The imports are separated in external dependencies (like std) and internal dependencies (like ascii and RequestError). To keep this separation you'll need to move the std imports above.

File: micro_http/src/common/headers.rs [hunk 3]
```diff
@@ -13,39 +17,178 @@ pub enum Header {
...
+            Err(RequestError::InvalidHeader)
         }
     }
 }
 
 /// Wrapper over the list of headers associated with a Request/Response.
 #[derive(Debug)]
 pub struct Headers {
-    headers: HashMap<Header, String>,
+    optional_headers: HashMap<Header, String>,
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: How about a different approach here. You just keep all headers in the HashMap and you add accessor functions for the headers you are interested in?

I am not expecting the accessors to be called so often that it will cause a major performance penalty. I find this approach to be cleaner and easier to extend in the future if needed.

```rust
pub struct Headers {
    headers: HashMap<Header, String>,
}

pub content_length(&self) -> i32 {
    match self.headers.get(Header::ContentLength) {
        // This unwrap is safe because you checked that ContentLength has a valid value when you parsed it.
        Some(value) => std::str::FromStr::from_str(value).unwrap(),
        None => 0,
    }
}

.....
```

Comment (andreeaflorescu) [reply micro_http/src/common/headers.rs]: I just noticed that you have accessors below. You would just need to change them a bit.

File: micro_http/src/common/headers.rs [hunk 4]
```diff
@@ -13,39 +17,178 @@ pub enum Header {
...
+        Err(RequestError::InvalidRequest)
+    }
+
+    /// Returns the content length of the body.
+    pub fn get_content_length(&self) -> i32 {
+        self.content_length
+    }
+
+    /// Returns `true` if the transfer encoding is chunked.
+    pub fn is_chunked(&self) -> bool {
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: nit: rename this to `chunked` because it is an accessor.

File: micro_http/src/common/headers.rs [hunk 5]
```diff
@@ -13,39 +17,178 @@ pub enum Header {
...
+                } else {
+                    return Ok(Some(body_string));
+                }
+            }
+        }
+        Err(RequestError::InvalidRequest)
+    }
+
+    /// Returns the content length of the body.
+    pub fn get_content_length(&self) -> i32 {
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: This should be named `content_length`.

File: micro_http/src/common/headers.rs [hunk 6]
```diff
@@ -13,39 +17,178 @@ pub enum Header {
...
+        self.content_length
+    }
+
+    /// Returns `true` if the transfer encoding is chunked.
+    pub fn is_chunked(&self) -> bool {
+        self.chunked
+    }
+
+    /// Returns `true` if the client is expecting the code 100.
+    pub fn is_expecting(&self) -> bool {
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: This should be named `expect`.

File: micro_http/src/request.rs [hunk 7]
```diff
@@ -152,19 +152,25 @@ impl<'a> Request<'a> {
     /// ```
     pub fn try_from(byte_stream: &'a [u8]) -> Result<Self, RequestError> {
         // The first line of the request is the Request Line. The line ending is LF.
-        let (request_line, _) = split(byte_stream, LF);
+        let (request_line, rest) = split(byte_stream, LF);
```
Comment (andreeaflorescu) [inline micro_http/src/request.rs]: Let's follow the HTTP protocol when splitting this.
Specifically when splitting a request by `LF` you end up with a `request_line` and another byte stream representing the `headers_and_body`.

To get all the headers we need a new split by CRLF. Only after you parsed both the request line and headers you proceed in creating a Request object.

```rust
    pub fn try_from(byte_stream: &'a [u8]) -> Result<Self, RequestError> {
        // The first line of the request is the Request Line. The line ending is LF.
        let (request_line, headers_and_body) = split(byte_stream, LF);
        if request_line.len() < RequestLine::min_len() {
            return Err(RequestError::InvalidRequest);
        }

        // The Request Line should include the trailing LF.
        let request_line = RequestLine::try_from(&byte_stream[..=request_line.len()])?;

        // Now let's split `headers_and_body` by LF to get only the headers so we can parse them.
        // We don't care about the body for now.
        // The split function needs to be updated so that you can split by CRLF. This is not yet
        // supported as far as I remember.
        let (headers, _) = split(headers_and_body, LF);
        let headers = Headers::try_from(headers)?;

         Ok(Request {
            request_line,
            headers,
            body: None,
        })
    }
```

File: micro_http/src/common/headers.rs [hunk 8]
```diff
@@ -13,39 +17,178 @@ pub enum Header {
...
 
     /// Adds a new header to the list.
     pub fn add(&mut self, header: Header, value: String) {
-        self.headers.insert(header, value);
+        self.optional_headers.insert(header, value);
+    }
+
+    /// Expects one header line and parses it, updating the header structure or returning an
+    /// error if the header is invalid.
+    pub fn parse_header_line(&mut self, header_line: &[u8]) -> Result<bool, RequestError> {
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: It's weird for `parse_header_line` to return a Result of bool. If you update the code as per the suggestion in `src/request.rs`, this will not be needed. Let's discuss about this offline.

File: micro_http/src/common/headers.rs [hunk 9]
```diff
@@ -13,39 +17,178 @@ pub enum Header {
...
+            let indexed_lines = text.split("\n").enumerate();
+            for (i, line) in indexed_lines {
+                if (line.len() == 0 || line == "\r") && !header_end {
+                    header_end = true;
+                    header_end_index = i;
+                } else {
+                    if !header_end {
+                        if let Err(_) = self.parse_header_line(line.as_bytes()) {
+                            error_occured = true;
+                        }
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: On error we should return RequestError.

File: micro_http/src/request.rs (hunk 7 above)
Comment (andreeaflorescu) [reply micro_http/src/request.rs]: As per our offline conversation this code will be changed anyway because we need to support streaming which will be implementing as part of the RequestConnection work.

File: micro_http/src/request.rs [hunk 10]
```diff
@@ -178,11 +184,33 @@ impl<'a> Request<'a> {
     pub fn http_version(&self) -> Version {
         self.request_line.http_version
     }
+
+    /// Parses all the fields found in `bytes`, assuming the string starts right
+    /// after the request line.
+    pub fn parse_headers(&mut self, bytes: &[u8]) -> Result<bool, RequestError> {
```
Comment (andreeaflorescu) [inline micro_http/src/request.rs]: As per our offline conversation we decided to go against setting the body here. The body will be setted in a higher level function depending on the header "Content-Length" which would be parsed before.

File: micro_http/src/common/headers.rs [hunk 11]
```diff
@@ -13,63 +12,220 @@ pub enum Header {
...
+            expect: false,
+            chunked: false,
         }
     }
 
-    /// Adds a new header to the list.
-    pub fn add(&mut self, header: Header, value: String) {
-        self.headers.insert(header, value);
+    /// Expects one header line and parses it, updating the header structure or returning an
+    /// error if the header is invalid.
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: You should add an `#Errors` section and describe when is the function returning `InvalidHeader` and when is it returning `InvalidRequest`.

File: micro_http/src/common/headers.rs [hunk 12]
```diff
@@ -13,63 +12,220 @@ pub enum Header {
...
+    /// use micro_http::Headers;
+    ///
+    /// let request_headers = Headers::try_from(b"Content-Length: 55\r\n\r\n");
+    /// ```
+    pub fn try_from(bytes: &[u8]) -> Result<Headers, RequestError> {
+        // Headers must be ASCII, so also UTF-8 valid.
+        if let Ok(text) = std::str::from_utf8(bytes) {
+            let mut headers = Headers::default();
+
+            let indexed_lines = text.split("\r\n");
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: How about naming this `header_lines`? Then the following for would read like:
```rust
for header_line in header_lines {
    if header_line.is_empty().....
}
```

File: micro_http/src/common/headers.rs [hunk 13]
```diff
@@ -13,63 +12,220 @@ pub enum Header {
...
+            for line in indexed_lines {
+                if line.is_empty() {
+                    break;
+                }
+                if let Err(RequestError::InvalidRequest) =
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: The following lines (183-188) can be simplified as follows:
```rust
headers.parse_header_line(line.as_bytes())?;
```

Read more about this simplification [here](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html).

File: micro_http/src/common/mod.rs [hunk 14]
```diff
@@ -99,7 +110,7 @@ impl Method {
 /// let version = Version::try_from(b"HTTP/1.1");
 /// assert!(version.is_ok());
 ///
-/// let version = Version::try_from(b"http/1.1");
+/// let version = Version::try_from(b"http/2.0");
```
Comment (andreeaflorescu) [inline micro_http/src/common/mod.rs]: Why is this changed? `http/1.1` should be invalid because the protocol name should be in uppercase.

File: micro_http/src/request.rs [hunk 15]
```diff
@@ -92,9 +98,16 @@ impl<'a> RequestLine<'a> {
...
+        let method_end = find(request_line, &[SP]);
+        let method = &request_line[..method_end];
+
+        let uri_and_version = &request_line[(method_end + 1)..];
+
+        let uri_end = find(uri_and_version, &[SP]);
+        let uri = &uri_and_version[..uri_end];
+
+        let version_end = find(uri_and_version, &[LF]);
+        let mut version = &uri_and_version[(uri_end + 1)..version_end];
```
Comment (andreeaflorescu) [inline micro_http/src/request.rs]: With these changes is the `remove_trailing_cr` method still needed?

File: micro_http/src/response.rs [hunk 16]
```diff
@@ -14,8 +14,12 @@ use headers::{Header, Headers, MediaType};
 #[allow(dead_code)]
 #[derive(Clone, Copy, PartialEq)]
 pub enum StatusCode {
-    /// 100, OK
+    /// 100, Continue
```
Comment (andreeaflorescu) [inline micro_http/src/response.rs]: You need to update the documentation of status code in `micro_http/src/lib.rs` and add the newly supported status codes.

File: micro_http/src/common/headers.rs (hunk 13 above)
Comment (georgepisaltu) [reply micro_http/src/common/headers.rs]: In the current design, we want to exit the function and return `InvalidRequest` (which invalidates the whole request) only when `parse_header_lines` returns that. Note that `parse_header_lines` can also return `InvalidHeader`, in which case the header line will be discarded but the request would still be valid.

The simplification with `?` would mean exiting the function even on invalid headers, which is not the desired outcome.

File: micro_http/src/request.rs (hunk 15 above)
Comment (georgepisaltu) [reply micro_http/src/request.rs]: It is still used once in parsing the request line. However, changes can be made in order to stop using it, the most obvious one would be to pass a slice without the CRLF sequence to the request line parser.

Comment (andreeaflorescu) [reply micro_http/src/request.rs]: Yes, that's what I was thinking about. Please add an issue to refactor this code as a next step if you don't plan to do it as part of the same PR.

File: micro_http/src/common/headers.rs [hunk 17]
```diff
@@ -13,63 +12,227 @@ pub enum Header {
...
         }
     }
 
-    /// Adds a new header to the list.
-    pub fn add(&mut self, header: Header, value: String) {
-        self.headers.insert(header, value);
+    /// Expects one header line and parses it, updating the header structure or returning an
+    /// error if the header is invalid.
+    ///
+    /// # Errors
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: Please also add an issue to fix the errors returned by micro_http as they're pretty much chaotic now.

`InvalidHeader` should be returned when the header is invalid and `Unsupported` should be returned when we found a header which we don't recognize and don't care about. The caller of `parse_header_line` should check the returned error and in case it is `Unsupported` the caller can ignore it, and when it is `InvalidHeader` it should return `BadRequest` or something similar and stop the parsing of the request.

File: micro_http/src/common/headers.rs [hunk 18]
```diff
@@ -13,63 +12,227 @@ pub enum Header {
...
+    /// `InvalidRequest` is returned when the parsed header suggests that the client is using
+    /// HTTP features that we do not support in this implementation, which invalidates the
+    /// request.
+    pub fn parse_header_line(&mut self, header_line: &[u8]) -> Result<(), RequestError> {
+        // Headers must be ASCII, so also UTF-8 valid.
+        match std::str::from_utf8(header_line) {
+            Ok(headers_str) => {
+                let entry = headers_str.split(": ").collect::<Vec<&str>>();
+                if entry.len() != 2 {
+                    return Err(RequestError::InvalidHeader);
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: This error shouldn't be be ignored. This should invalidate the request.

File: micro_http/src/common/headers.rs [hunk 19]
```diff
@@ -13,63 +12,227 @@ pub enum Header {
...
+                if let Ok(head) = Header::try_from(entry[0].as_bytes()) {
+                    match head {
+                        Header::ContentLength => {
+                            let try_numeric: Result<i32, std::num::ParseIntError> =
+                                std::str::FromStr::from_str(entry[1].trim());
+                            if try_numeric.is_ok() {
+                                self.content_length = try_numeric.unwrap();
+                                Ok(())
+                            } else {
+                                Err(RequestError::InvalidHeader)
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: Same as before.

File: micro_http/src/common/headers.rs [hunk 20]
```diff
@@ -13,63 +12,226 @@ pub enum Header {
...
-    /// Adds a new header to the list.
-    pub fn add(&mut self, header: Header, value: String) {
-        self.headers.insert(header, value);
+    /// Expects one header line and parses it, updating the header structure or returning an
+    /// error if the header is invalid.
+    ///
+    /// # Errors
+    /// `UnsupportedHeader` is returned when the parsed header line is not of interest
+    /// to us or when it is unrecognizable.
+    /// `InvalidHeader` is returned when the parsed header suggests that the client is using
```
Comment (andreeaflorescu) [inline micro_http/src/common/headers.rs]: nit: InvalidHeader is also returned when the header is malformed like for example when the header line does not contain 2 tokens separated by `:`.

File: micro_http/src/common/headers.rs [hunk 21]
```diff
@@ -13,63 +12,226 @@ pub enum Header {
...
         match self {
             Header::ContentLength => b"Content-Length",
             Header::ContentType => b"Content-Type",
+            Header::Expect => b"Expect",
+            Header::TransferEncoding => b"Transfer-Encoding",
+        }
+    }
+
+    fn try_from(string: &[u8]) -> Result<Self, RequestError> {
+        if let Ok(s) = String::from_utf8(string.to_vec()) {
```
Comment (serban300) [inline micro_http/src/common/headers.rs]: I would rename this variable to `utf8_string` or something else. `s` is not very suggestive.

File: micro_http/src/request.rs [hunk 22]
```diff
@@ -3,19 +3,25 @@
...
-// The separator is not included in the return values.
-fn split(bytes: &[u8], separator: u8) -> (&[u8], &[u8]) {
-    match bytes.iter().position(|byte| *byte == separator) {
-        Some(index) => (&bytes[..index], &bytes[index + 1..]),
-        None => (&[], bytes),
+// Helper functions used for parsing the HTTP Request.
+// Finds the first occurrence of the sequence in a byte slice and returns its
+// position in the slice.
+// If no occurrence is found, it returns `None`.
+fn find(bytes: &[u8], sequence: &[u8]) -> Option<usize> {
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply micro_http/src/request.rs]: The performance for this function is not critical, I will go with your second suggestion.
